### PR TITLE
MWPW-140481: Delayed modal

### PR DESCRIPTION
**Pull request by mirafedas**
This feature enables the delayed display of a modal, primarily intended for promotional purposes.

When parsing the promotional manifest, if we encounter a rule labeled insertContentAfter containing a link with valid parameters for delay and display, we dynamically generate a hidden link within the specified selector. This link behaves conventionally, triggering a modal upon click. Following this setup, we introduce a delay of some number of seconds before unveiling the modal, simultaneously appending a modal hash to the browser's URL. If the modal is configured for a one-time session display (display="session"), we remember this in sessionStorage to prevent subsequent displays upon page reload.

Resolves: [MWPW-140481](https://jira.corp.adobe.com/browse/MWPW-140481)

[Authoring documentation](https://main--milo--mirafedas.hlx.live/docs/authoring/promo/delayed-modal)

**Test URLs:**
- Before: not applicable, because this is a newly added feature
- After: 
[First page](https://mwpw-140481-delayed-modal--milo--mirafedas.hlx.live/drafts/mirafedas/showdelayedmodalhere?manifest=%2Ffragments%2Fmirafedas%2Fpromos%2F2024%2Fglobal%2Fdelayed-modal-promo%2Fdelayed-modal-promo.json--all)
[Second page](https://mwpw-140481-delayed-modal--milo--mirafedas.hlx.page/drafts/mirafedas/secondpage)
Feel free to change the `display` param to either `pageload` or `session` in [this file](https://adobe.sharepoint.com/:x:/r/sites/adobecom/_layouts/15/Doc.aspx?sourcedoc=%7B58183116-E8D6-42D1-A007-82226E8405CA%7D&file=delayed-modal-promo.xlsx&action=default&mobileredirect=true)

### CODE REVIEW

File: libs/features/personalization/personalization.js [hunk 1]
```diff
@@ -99,18 +99,105 @@ export const preloadManifests = ({ targetManifests = [], persManifests = [] }) =
 
 export const getFileName = (path) => path?.split('/').pop();
 
-const createFrag = (el, url, manifestId) => {
-  let href = url;
+export async function initDelayedModal({
```
Comment (yesil) [inline libs/features/personalization/personalization.js:102]: I think modal logic should go in modal.js

File: libs/features/personalization/personalization.js [hunk 2]
```diff
@@ -99,18 +99,105 @@ export const preloadManifests = ({ targetManifests = [], persManifests = [] }) =
...
+  loadLink(`${miloLibs}/blocks/modal/modal.css`, { rel: 'stylesheet' });
+  if (displayMode === DELAYED_MODAL_DISPLAY_MODE.oncePerPageLoad) {
+    setTimeout(() => {
+      getModal(details);
+    }, delayInMiliseconds);
+  } else if (displayMode === DELAYED_MODAL_DISPLAY_MODE.oncePerSession) {
+    if (!window.sessionStorage.getItem('wasDelayedModalShown')) {
+      setTimeout(() => {
+        getModal(details);
+        window.sessionStorage.setItem('wasDelayedModalShown', true);
```
Comment (yesil) [inline libs/features/personalization/personalization.js:124]: this implementation supports one single delayed modal.
if the user navigates from one page to another in the same session, I think the other modals will not be shown.

File: test/features/personalization/delayedModal.test.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  expect(a.getAttribute('data-modal-hash')).to.equal('#delayed-modal');
+  expect(a.getAttribute('data-modal-path')).to.equal('/testpage/');
+  expect(a.getAttribute('style')).to.equal('display: none');
+  expect(a.classList.contains('modal')).to.be.true;
+  expect(a.classList.contains('link-block')).to.be.true;
+  a.remove();
+});
+
+it('creates and opens the delayed modal', async () => {
+  const hash = '#delayed-modal';
```
Comment (yesil) [inline test/features/personalization/delayedModal.test.js:53]: `delayed-modal` shounds to technical.

I wonder if we could adopt a more short/cryptic naming here, like `md` or `dm`.

**Review by auniverseaway — CHANGES_REQUESTED** [review]
I think this can be simpler.

File: libs/features/personalization/personalization.js (hunk 2 above)
Comment (mirafedas) [reply libs/features/personalization/personalization.js]: @yesil 
I created two test pages to verify this:
1. [First page](https://mwpw-140481-delayed-modal--milo--mirafedas.hlx.live/drafts/mirafedas/showdelayedmodalhere?manifest=%2Ffragments%2Fmirafedas%2Fpromos%2F2024%2Fglobal%2Fdelayed-modal-promo%2Fdelayed-modal-promo.json--all)
2. [Second page](https://mwpw-140481-delayed-modal--milo--mirafedas.hlx.page/drafts/mirafedas/secondpage)
On both pages I placed some traditional modals.
To me all works as expected.
If you want to test yourself, feel free to change the `display` param to either `pageload` or `session` in [this file](https://adobe.sharepoint.com/:x:/r/sites/adobecom/_layouts/15/Doc.aspx?sourcedoc=%7B58183116-E8D6-42D1-A007-82226E8405CA%7D&file=delayed-modal-promo.xlsx&action=default&mobileredirect=true)

Comment (yesil) [reply libs/features/personalization/personalization.js]: this test doesn't cover oncePerSession option.
please use oncePerSession, and link one page to the other one and try to navigate.
I expect the second modal will not be shown.

**Review by chrischrischris — CHANGES_REQUESTED** [review]
I haven't had a chance for an in-depth review but wanted to ask - why are all the changes in personalization needed?

Also we don't want personalization.js to load any external files (modal.js) unless absolutely needed.

**Comment by chrischrischris** [comment]
My thoughts:
* No modal code should be in personalization.js, all functionality should be in modal.js
* Instead of query params, we should have the params built into the modal name: `#modal-name:delay=3:display=session`.
  * This allows for the modal.js code to remove the  params from the url without needing to do a refresh.
* delayed modals should work without MEP - e.g: www.adobe.com/nz/mypage#modal-name:delay=3:display=session will work on page load.

With these changes, there are no changes needed to MEP.  The fragment link set in the manifest would contain the info and modal.js would handle everything else.

More minor thoughts:
* What should the behavior be if multiple fragments specify a delayed modal?
* What happens when a modal is deep linked (so displaying immediately on page load) and a fragment has a delayed modal?
* sessionstorage should differentiate between what delayed modal was shown.  A single `wasDelayedModalShown` key would stop a possible other delayed modal unrelated to the original one from being show.  Something like `shown:modal-name`.

**Comment by Roycethan** [comment]
@mirafedas  I don't see modal loading on successive page loads, Repro: First load https://mwpw-140481-delayed-modal--milo--mirafedas.hlx.live/drafts/mirafedas/showdelayedmodalhere?manifest=%2Ffragments%2Fmirafedas%2Fpromos%2F2024%2Fglobal%2Fdelayed-modal-promo%2Fdelayed-modal-promo.json--all#delayed-modal then Refresh the browser and check to see if it load

**Comment by mirafedas** [comment]
Closing this PR as these changes are not actual anymore, and opening a new one

**Pull request closed without merging**